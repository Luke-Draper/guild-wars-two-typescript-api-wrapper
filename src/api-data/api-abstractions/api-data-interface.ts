import KeyValueInterface from "./key-val-interface";

/**
 * An interface representing all nodes (incuding root) obtained from the API
 */
export abstract class APINodeInterface {
	/**
	 * The path extension from root (<a>https://api.guildwars2.com/v2</a>)to access this data<br>
	 * @example <br><br>
	 * The path extension at "<a>https://api.guildwars2.com/v2/items/357</a>" is "/items/357"
	 */
	pathExtension: string;
	/**
	 * The data accessed by getting this API endpoint<br>
	 * @example <br><br>
	 * The data returned from "<a>https://api.guildwars2.com/v2/items/357</a>" is
	 * <br>{
	 * <br>&nbsp; "name": "Mighty Magician Coat of Vampirism",
	 * <br>&nbsp; "description": "",
	 * <br>&nbsp; "type": "Armor",
	 * <br>&nbsp; "level": 20,
	 * <br>&nbsp; "rarity": "Masterwork",
	 * <br>&nbsp; "vendor_value": 72,
	 * <br>&nbsp; "default_skin": 37,
	 * <br>&nbsp; "game_types": [
	 * <br>&nbsp; &nbsp; "Activity",
	 * <br>&nbsp; &nbsp; "Wvw",
	 * <br>&nbsp; &nbsp; "Dungeon",
	 * <br>&nbsp; &nbsp; "Pve"
	 * <br>&nbsp; ],
	 * <br>&nbsp; "flags": [
	 * <br>&nbsp; &nbsp; "SoulBindOnUse"
	 * <br>&nbsp; ],
	 * <br>&nbsp; "restrictions": [],
	 * <br>&nbsp; "id": 357,
	 * <br>&nbsp; "chat_link": "[&AgFlAQAA]",
	 * <br>&nbsp; "icon": "https://render.guildwars2.com/file/D1BBAB94232E2AD1F1253D573D45C4DCECDF0E02/61018.png",
	 * <br>&nbsp; "details": {
	 * <br>&nbsp; &nbsp; "type": "Coat",
	 * <br>&nbsp; &nbsp; "weight_class": "Light",
	 * <br>&nbsp; &nbsp; "defense": 59,
	 * <br>&nbsp; &nbsp; "infusion_slots": [],
	 * <br>&nbsp; &nbsp; "infix_upgrade": {
	 * <br>&nbsp; &nbsp; &nbsp; "id": 137,
	 * <br>&nbsp; &nbsp; &nbsp; "attributes": [
	 * <br>&nbsp; &nbsp; &nbsp; &nbsp; {
	 * <br>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "attribute": "Power",
	 * <br>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "modifier": 19
	 * <br>&nbsp; &nbsp; &nbsp; &nbsp; }
	 * <br>&nbsp; &nbsp; &nbsp; ]
	 * <br>&nbsp; &nbsp; },
	 * <br>&nbsp; &nbsp; "suffix_item_id": 24709,
	 * <br>&nbsp; &nbsp; "secondary_suffix_item_id": ""
	 * <br>&nbsp; }
	 * <br>}
	 */
	rawData: object;
	/**
	 * This should be false until this.setupDataStructure() is complete
	 */
	setupComplete: boolean;
	/**
	 * This function takes the raw data from the endpoint and parses it into a usable data structure
	 * @param pathExtension
	 * The path extension of this endpoint
	 * @param rawData
	 * The raw data obtained from this endpoint
	 * @returns
	 * Should return this
	 */
	constructor(pathExtension: string, rawData: object) {
		this.pathExtension = pathExtension;
		this.rawData = rawData;
		this.setupComplete = false;
	}
	/**
	 * This function should be called once all of the data from the API has been stored.<br>
	 * It should link this data structure with the rest of the data structures
	 * @returns
	 * Should return this
	 */
	abstract setupDataStructure(): APINodeInterface;
}

/**
 * An interface representing all data obtained from the API
 */
export abstract class APIDataInterface extends APINodeInterface {
	/**
	 * The parent api data endpoint
	 */
	parent: APINodeInterface;
	/**
	 * This function takes the raw data from the endpoint and parses it into a usable data structure
	 * @param pathExtension
	 * The path extension of this endpoint
	 * @param rawData
	 * The raw data obtained from this endpoint
	 * @returns
	 * Should return this
	 */
	constructor(
		parent: APINodeInterface,
		pathExtension: string,
		rawData: object
	) {
		super(pathExtension, rawData);
		this.parent = parent;
	}
	/**
	 * This function should be called once all of the data from the API has been stored.<br>
	 * It should link this data structure with the rest of the data structures
	 * @returns
	 * Should return this
	 */
	abstract setupDataStructure(): APIDataInterface;
}

/**
 * An interface representing all data obtained from an element endpoint of the API<br>
 * @example <br><br>
 * "<a>https://api.guildwars2.com/v2/items/357</a>" is an element because it has an id identifying it in the endpoint "/items" list
 */
export abstract class APIElementDataInterface extends APIDataInterface {
	/**
	 * The id of this data element
	 */
	abstract id: number;
	/**
	 * This function should be called once all of the data from the API has been stored.<br>
	 * It should link this data structure with the rest of the data structures
	 * @returns
	 * Should return this
	 */
	abstract setupDataStructure(): APIElementDataInterface;
}

/**
 * An interface representing all endpoints containing a list of ids pointing to element endpoints of the API<br>
 * @example <br><br>
 * "<a>https://api.guildwars2.com/v2/items</a>" is a list because it has an list of ids identifying sub endpoints "/items/<id>"
 */
export abstract class APIListDataInterface<
	Element extends APIElementDataInterface
> extends APIDataInterface {
	/**
	 * The list of ids referencing element endpoints
	 */
	abstract ids: number[];
	abstract idsAllAvailable: boolean;
	abstract elements: Element[];
	abstract setupElements(elements: Element[]): void;
	/**
	 * This function should be called once all of the data from the API has been stored.<br>
	 * It should link this data structure with the rest of the data structures
	 * @returns
	 * Should return this
	 */
	abstract setupDataStructure(): APIListDataInterface<Element>;
}

export abstract class APISortedListDataInterface<
	Element extends APIElementDataInterface
> extends APIListDataInterface<Element> {
	abstract elementIdToIndexDict: KeyValueInterface;
	abstract elementNameToIndexDict: KeyValueInterface;
	abstract setupSortedElements(elements: Element[]): void;
	abstract getElementById(id: number | string): void;
	abstract getElementByName(name: string): void;
	/**
	 * This function should be called once all of the data from the API has been stored.<br>
	 * It should link this data structure with the rest of the data structures
	 * @returns
	 * Should return this
	 */
	abstract setupDataStructure(): APISortedListDataInterface<Element>;
}

export abstract class APIDataParentInterface<
	Child extends APIDataInterface
> extends APIDataInterface {
	abstract children: Child[];
	abstract setupChildren(children?: Child[]): void;
	/**
	 * This function should be called once all of the data from the API has been stored.<br>
	 * It should link this data structure with the rest of the data structures
	 * @returns
	 * Should return this
	 */
	abstract setupDataStructure(): APIDataParentInterface<Child>;
}

export abstract class APIDataPointFromInterface extends APIDataInterface {
	abstract pointedTo: APIDataPointedToInterface[];
	abstract addPointTo(pointTo: APIDataPointedToInterface): void;
	/**
	 * This function should be called once all of the data from the API has been stored.<br>
	 * It should link this data structure with the rest of the data structures
	 * @returns
	 * Should return this
	 */
	abstract setupDataStructure(): APIDataPointFromInterface;
}

export abstract class APIDataPointedToInterface extends APIDataInterface {
	abstract pointedFrom: APIDataPointFromInterface[];
	abstract addPointFrom(pointFrom: APIDataPointFromInterface): void;
	/**
	 * This function should be called once all of the data from the API has been stored.<br>
	 * It should link this data structure with the rest of the data structures
	 * @returns
	 * Should return this
	 */
	abstract setupDataStructure(): APIDataPointedToInterface;
}
