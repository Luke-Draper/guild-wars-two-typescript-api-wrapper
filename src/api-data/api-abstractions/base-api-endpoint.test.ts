import BaseAPIEndpoint from "./base-api-endpoint";
import { APIDataInterface, APINodeInterface } from "./api-data-interface";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";

describe("APIEndpoint", () => {
	let mockAdapter: MockAdapter;
	class MockAPIData extends APIDataInterface {
		setupDataStructure(): APIDataInterface {
			return this;
		}
	}
	class MockAPINode extends APINodeInterface {
		setupDataStructure(): APINodeInterface {
			return this;
		}
	}
	const mockRoot = new MockAPINode("https://api.guildwars2.com/v2", {});

	beforeEach(() => {
		mockAdapter = new MockAdapter(axios);
		mockAdapter
			.onGet("https://api.guildwars2.com/v2/test", {
				params: {}
			})
			.reply(200, { location: "test_endpoint" });
		mockAdapter
			.onGet("https://api.guildwars2.com/v2/test", {
				params: { paramaters: [1, 2, 3] }
			})
			.reply(200, { location: "test_endpoint_with_params" });
		mockAdapter
			.onGet("https://api.guildwars2.com/v2/test", {
				params: { access_token: "test_token", paramaters: [1, 2, 3] }
			})
			.reply(200, { location: "test_endpoint_with_authorization" });
	});

	it("instantiates", () => {
		expect(
			new BaseAPIEndpoint<MockAPIData>(
				"/test",
				false,
				false,
				(parent: MockAPINode, pathExtension: string, rawData: object) => {
					return new MockAPIData(parent, pathExtension, rawData);
				},
				mockRoot
			)
		).toBeInstanceOf(BaseAPIEndpoint);
	});
	it("instantiates with authentication", () => {
		expect(
			new BaseAPIEndpoint<MockAPIData>(
				"/test",
				false,
				true,
				(parent: MockAPINode, pathExtension: string, rawData: object) => {
					return new MockAPIData(parent, pathExtension, rawData);
				},
				mockRoot,
				{ access_token: "test_token" }
			)
		).toBeInstanceOf(BaseAPIEndpoint);
	});
	it("doesn't instantiate authenticated endpoint without authentication", () => {
		expect(() => {
			let test = new BaseAPIEndpoint<MockAPIData>(
				"/test",
				false,
				true,
				(parent: MockAPINode, pathExtension: string, rawData: object) => {
					return new MockAPIData(parent, pathExtension, rawData);
				},
				mockRoot,
				{ not_access_token: "test_token" }
			);
		}).toThrowErrorMatchingSnapshot();
	});

	it("allows data promise to be gotten", () => {
		expect.assertions(1);
		let mock = new BaseAPIEndpoint<MockAPIData>(
			"/test",
			true,
			false,
			(parent: MockAPINode, pathExtension: string, rawData: object) => {
				return new MockAPIData(parent, pathExtension, rawData);
			},
			mockRoot
		);
		return mock.getDataPromise().then(data => {
			expect(data).toBeInstanceOf(MockAPIData);
		});
	});

	it("sends a request to the proper endpoint", () => {
		expect.assertions(2);
		let mock = new BaseAPIEndpoint<MockAPIData>(
			"/test",
			false,
			false,
			(parent: MockAPINode, pathExtension: string, rawData: object) => {
				return new MockAPIData(parent, pathExtension, rawData);
			},
			mockRoot
		);
		return mock.requestEndpoint().then(result => {
			expect(result).not.toEqual(undefined);
			expect(result).toEqual({
				location: "test_endpoint"
			});
		});
	});
	it("sends a request to the proper endpoint with paramaters", () => {
		expect.assertions(2);
		let mock = new BaseAPIEndpoint<MockAPIData>(
			"/test",
			false,
			false,
			(parent: MockAPINode, pathExtension: string, rawData: object) => {
				return new MockAPIData(parent, pathExtension, rawData);
			},
			mockRoot,
			{ paramaters: [1, 2, 3] }
		);
		return mock.requestEndpoint().then(result => {
			expect(result).not.toEqual(undefined);
			expect(result).toEqual({
				location: "test_endpoint_with_params"
			});
		});
	});
	it("sends a request to the proper endpoint with authorization", () => {
		expect.assertions(2);
		let mock = new BaseAPIEndpoint<MockAPIData>(
			"/test",
			false,
			false,
			(parent: MockAPINode, pathExtension: string, rawData: object) => {
				return new MockAPIData(parent, pathExtension, rawData);
			},
			mockRoot,
			{
				access_token: "test_token",
				paramaters: [1, 2, 3]
			}
		);
		return mock.requestEndpoint().then(result => {
			expect(result).not.toEqual(undefined);
			expect(result).toEqual({
				location: "test_endpoint_with_authorization"
			});
		});
	});

	it("sets up without sending a request", () => {
		expect.assertions(2);
		let mock = new BaseAPIEndpoint<MockAPIData>(
			"/test",
			false,
			false,
			(parent: MockAPINode, pathExtension: string, rawData: object) => {
				return new MockAPIData(parent, pathExtension, rawData);
			},
			mockRoot
		);
		return mock.setupData(false).then(result => {
			expect(result).not.toEqual(undefined);
			expect(result.rawData).toEqual({});
		});
	});

	it("sets up and sends a request to the proper endpoint", () => {
		expect.assertions(2);
		let mock = new BaseAPIEndpoint<MockAPIData>(
			"/test",
			false,
			false,
			(parent: MockAPINode, pathExtension: string, rawData: object) => {
				return new MockAPIData(parent, pathExtension, rawData);
			},
			mockRoot
		);
		return mock.setupData(true).then(result => {
			expect(result).not.toEqual(undefined);
			expect(result.rawData).toEqual({
				location: "test_endpoint"
			});
		});
	});
	it("sets up and sends a request to the proper endpoint with paramaters", () => {
		expect.assertions(2);
		let mock = new BaseAPIEndpoint<MockAPIData>(
			"/test",
			false,
			false,
			(parent: MockAPINode, pathExtension: string, rawData: object) => {
				return new MockAPIData(parent, pathExtension, rawData);
			},
			mockRoot,
			{ paramaters: [1, 2, 3] }
		);
		return mock.setupData(true).then(result => {
			expect(result).not.toEqual(undefined);
			expect(result.rawData).toEqual({
				location: "test_endpoint_with_params"
			});
		});
	});
	it("sets up and sends a request to the proper endpoint with authorization", () => {
		expect.assertions(2);
		let mock = new BaseAPIEndpoint<MockAPIData>(
			"/test",
			false,
			false,
			(parent: MockAPINode, pathExtension: string, rawData: object) => {
				return new MockAPIData(parent, pathExtension, rawData);
			},
			mockRoot,
			{
				access_token: "test_token",
				paramaters: [1, 2, 3]
			}
		);
		return mock.setupData(true).then(result => {
			expect(result).not.toEqual(undefined);
			expect(result.rawData).toEqual({
				location: "test_endpoint_with_authorization"
			});
		});
	});
});
