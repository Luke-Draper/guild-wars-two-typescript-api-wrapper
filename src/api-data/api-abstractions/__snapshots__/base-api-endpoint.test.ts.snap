// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIEndpoint doesn't instantiate authenticated endpoint without authentication 1`] = `"Unable to access authenticated endpoint without an access token"`;
